src/sbpir/mod.rs

```diff
@@ -331,31 +386,88 @@ impl<F> StepType<F> {
         signal
     }
 
-    fn add_internal(&mut self, signal: InternalSignal) {
+    pub fn add_internal(&mut self, signal: InternalSignal) {
         self.annotations
             .insert(signal.uuid(), signal.annotation.to_string());
         self.signals.push(signal);
     }
 
-    pub fn add_constr(&mut self, annotation: String, expr: PIR<F>) {
+    pub fn add_constr(&mut self, annotation: String, expr: PIR<F, M>) {
         let condition = Constraint { annotation, expr };
 
         self.constraints.push(condition)
     }
 
-    pub fn add_transition(&mut self, annotation: String, expr: PIR<F>) {
+    pub fn add_transition(&mut self, annotation: String, expr: PIR<F, M>) {
         let condition = TransitionConstraint { annotation, expr };
 
         self.transition_constraints.push(condition)
     }
 }
 
-impl<F: Clone + Eq + Hash> StepType<F> {
-    pub fn decomp_constraints<M>(&mut self, mut decomposer: M)
+impl<F: Field + Hash, M: Clone> StepType<F, M> {
+    pub fn with_meta<N: Clone, ApplyMetaFn>(&self, apply_meta: ApplyMetaFn) -> StepType<F, N>
+    where
+        ApplyMetaFn: Fn(&Expr<F, Queriable<F>, ()>) -> N + Clone,
```

Comment:
Yes I agree, at some point I did it but I don't know why I moved back again, and I think I thought it would be safer forcing to remove the meta first but I think is not so relevant actually. Thank you for pointing that